File: leetcode/23.merge-k-sorted-lists-iteration.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        int n = lists.length;
        if(n==0) return null;
        for(int len=1;len<n;len*=2) 
            for(int j=0;j+len<n;j+=2*len) {
                lists[j]=merge(lists[j], lists[j+len]);
            }
        return lists[0];
    }
    
    ListNode merge(ListNode a, ListNode b) {
        ListNode head=new ListNode(0);
        ListNode p = head;
        while(a!=null && b!=null) {
            if(a.val<b.val) {
                p.next=a;
                a=a.next;
            } else {
                p.next=b;
                b=b.next;
            }
            p=p.next;
        }
        
        if(a!=null) {
            p.next=a;
        } else if(b!=null) {
            p.next=b;
        }
        
        return head.next;
    }
}

File: leetcode/23.merge-k-sorted-lists-iteration.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        def merge(a, b):
            dummy = ListNode(0)
            last = dummy
            while a and b:
                if a.val<b.val:
                    last.next=a
                    a=a.next
                else:
                    last.next=b
                    b=b.next
                last=last.next
            if a:
                last.next=a
            elif b:
                last.next=b
            return dummy.next
        if not lists: return None
        l = 1
        while l<len(lists):
            j = 0
            while j+l<len(lists):
                lists[j]=merge(lists[j],lists[j+l])
                j+=2*l
            l+=l
        return lists[0]
